application/src/Serde.tsx

```diff
@@ -1,480 +1,494 @@
   }
   return {
-    serializer:serializer,
-    deserializer:deserializer,
-  }
+    serializer,
+    deserializer,
+  };
 }
 
-type Serializer = (data:any) => Uint8Array;
+type Serializer = (seriailzedData: BitArray, data: any) => void;
 // return field name, and its appropriate Serializer
-type RecordSerializer = Record<string,Serializer>;
+type RecordSerializer = Record<string, Serializer>;
 
-// returns the deserializeData, and the number of consumedBytes
-type Deserializer = (data:Uint8Array)=>[any, number];
+// returns the deserializedData
+type Deserializer = (serializedData: BitArray) => any;
 // return field name, and its appropriate Deserializer
 type RecordDeserializer = Record<string, Deserializer>;
 
-type RecordFieldsSerde = {"serializer":RecordSerializer,"deserializer":RecordDeserializer};
-type ArrayItemsSerde={"serializer":Serializer,"deserializer":Deserializer};
-
+type FieldsRecordSerde = {
+  serializer: RecordSerializer;
+  deserializer: RecordDeserializer;
+};
+interface Serde{
+  serializer: Serializer;
+  deserializer: Deserializer;
+}
 
-function serdeOptionalFieldsRecord(fieldsSerde:RecordFieldsSerde){
-  function fieldsExistenceByteCount(fields:RecordSerializer|RecordDeserializer):number{
-    return Math.ceil((Object.keys(fields).length)/8);
-  }
-  function serializer(fieldsSerializers:RecordSerializer,data:Record<string,any>):Uint8Array{
-    let fieldsExistenceValues = 0;
-    let serializedData:Uint8Array =  new Uint8Array(0);
-    for (const fieldIndex in Object.keys(fieldsSerializers)){
-      let field=Object.keys(fieldsSerializers)[fieldIndex];
-      fieldsExistenceValues=fieldsExistenceValues << 1;
-      try{
-        let fieldValue=Reflect.get(data,field);
-        if (fieldValue==undefined){
+function serdeOptionalFieldsRecord(fieldsSerde: FieldsRecordSerde) :Serde {
+  function fieldsExistenceBitCount(
+    fields: RecordSerializer | RecordDeserializer
+  ): number {
+    return Object.keys(fields).length;
+  }
+  function serializer(
+    serializedData: BitArray,
+    fieldsSerializers: RecordSerializer,
+    data: Record<string, any>
+  ) {
+    let fieldsExistenceArray = new BitArray();
+    let serializedFields = new BitArray();
+
+    for (const fieldIndex in Object.keys(fieldsSerializers)) {
+      let field = Object.keys(fieldsSerializers)[fieldIndex];
+      try {
+        let fieldValue = Reflect.get(data, field);
+        if (fieldValue == undefined) {
+          fieldsExistenceArray.insertBool(false);
           continue;
         }
-        let serializedField = fieldsSerializers[field](fieldValue);
-        serializedData=combineByteArrays(serializedData,serializedField);
-        fieldsExistenceValues=fieldsExistenceValues |= 1;
-      }catch (error){}
+        fieldsExistenceArray.insertBool(true);
+        fieldsSerializers[field](serializedFields, fieldValue);
+      } catch (error) {
+        fieldsExistenceArray.insertBool(false);
+      }
```

Comment:
why try catch??

Reply:
if the field does not exist trying to get it will crash the program, we do not want to do that and instead want to detect that the field does not exist